Add more smart pointers to Length class

abigailfox opened the pull request:
#### 532ff80625a709ef09dead4714e87f73c86f8a57
<pre>
Add more smart pointers to Length class
<a href="https://bugs.webkit.org/show_bug.cgi?id=274919">https://bugs.webkit.org/show_bug.cgi?id=274919</a>
<a href="https://rdar.apple.com/129022845">rdar://129022845</a>

Reviewed by Chris Dumez.

* Source/WebCore/platform/Length.cpp:
(WebCore::Length::protectedCalculationValue const):
(WebCore::Length::nonNanCalculatedValue const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/Length.h:

Canonical link: <a href="https://commits.webkit.org/279608@main">https://commits.webkit.org/279608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/313201e1971c0839ab00e8cb5b7befbd942f1281

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33202 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57112 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4557 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4418 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43593 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2989 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55932 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31410 "Found 1 new test failure: compositing/overflow/clipping-ancestor-with-accelerated-scrolling-ancestor.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24733 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28238 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3879 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49934 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4070 "Found 1 new test failure: imported/w3c/web-platform-tests/resource-timing/status-codes-create-entry.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58707 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4128 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51004 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30196 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46712 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50344 "Found 3 new API test failures: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11753 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31132 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29975 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -133,7 +133,7 @@ class CalculationValueMap {
     struct Entry {
         uint64_t referenceCountMinusOne;
-        CalculationValue* value;
+        RefPtr<CalculationValue> value;
```
If we do this, it seems we need to update `insert()` / `deref()` since they currently leakRef to make sure the pointer stays valid and adopt to delete. Seems we wouldn't need this logic anymore if we just ref in the Entry.

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -133,9 +133,10 @@ class CalculationValueMap {
-        CalculationValue* value;
+        RefPtr<CalculationValue> value;
         Entry();
         Entry(CalculationValue&);
```
I do not believe we still need this.

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -154,6 +155,12 @@ inline CalculationValueMap::Entry::Entry(CalculationValue& value)
 }
 
+inline CalculationValueMap::Entry::Entry(Ref<CalculationValue>&& value)
+    : referenceCountMinusOne(0)
```
Would be nice to initialize this inline so the constructors don't have to.

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -163,12 +170,11 @@ inline unsigned CalculationValueMap::insert(Ref<CalculationValue>&& value)
 
-    // The leakRef below is balanced by the adoptRef in the deref member function.
-    Entry leakedValue = value.leakRef();
+    Entry entry(std::forward<Ref<CalculationValue>>(value));
```
No, you want `WTFMove()` here, not forward.

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -132,25 +132,23 @@ class CalculationValueMap {
 inline CalculationValueMap::Entry::Entry()
-    : referenceCountMinusOne(0)
-    , value(nullptr)
+    : value(nullptr)
```
Not needed.

cdumez commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -132,25 +132,23 @@ class CalculationValueMap {
     HashMap<unsigned, Entry> m_map;
 };
 
 inline CalculationValueMap::Entry::Entry()
```
We don't need this whole constructor here. we can just use:
```
Entry() = default;
```

above.

abigailfox committed:
Add more smart pointers to Length class
https://bugs.webkit.org/show_bug.cgi?id=274919
rdar://129022845

Reviewed by Chris Dumez.

* Source/WebCore/platform/Length.cpp:
(WebCore::Length::protectedCalculationValue const):
(WebCore::Length::nonNanCalculatedValue const):
(WebCore::operator<<):
* Source/WebCore/platform/Length.h:

Canonical link: https://commits.webkit.org/279608@main
